### PR TITLE
Unmatched requests

### PR DESCRIPTION
**Pull request by TristanSpeakEasy**
This PR contains a few things in an attempt to bring it inline with the Go SDK:

- Fixes precision issues with capturing time
- Adds support for collecting and setting path hints
- Requires the setting of an appID and version ID, while making it possible to configure the SDK globally or per route.
- Adds support for collecting a customerID to help with the embed project. (Go SDK doesn't do this yet but will be updated shortly)
- Adds testing for the interceptor
- Replaces local grpc compiling for pulling grpc schemas from an external library generated from our `speakeasy-schemas` repo

- [x] TODO: Will need to merge after @alexadrake as I have also made some changes to the har-java lib and updated the binary file here https://github.com/speakeasy-api/har-java/pull/2/files

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+
+Speakeasy is your API Platform team as a service. Use our drop in SDK to manage all your API Operations including
+customer facing usage embeds, test case generation from traffic, and understanding API drift.
+
+This is the Speakeasy JVM SDK for evaluating API requests/responses. Currently only supports the Springboot API
+Framework. If you are integrating for the first time please start at the [Quickstart](https://docs.speakeasyapi.dev/overview/quickstart) on our docs home.
+
+## Requirements
+
+- Spring Boot 2.4.X
```
Comment (alexadrake) [inline README.md:13]: We're supporting in 2.7.x

File: src/main/java/dev/speakeasyapi/sdk/utils/SpeakeasyInterceptor.java [hunk 2]
```diff
@@ -14,25 +23,105 @@
...
+    public SpeakeasyInterceptor(String apiKey, String apiID, String versionID, Logger logger) {
+        init(apiKey, apiID, versionID, logger);
+    }
+
+    public SpeakeasyInterceptor(SpeakeasyRequestResponseHandler requestResponseHandler) {
+        this.requestResponseHandler = requestResponseHandler;
+    }
+
+    private void init(String apiKey, String apiID, String versionID, Logger logger) {
+        String serverURL = System.getenv("SPEAKEASY_SERVER_URL");
```
Comment (alexadrake) [inline src/main/java/dev/speakeasyapi/sdk/utils/SpeakeasyInterceptor.java:48]: I think these  (server url and secure grpc) should be kept as `application.properties` to keep our patterns more idiomatically java.

File: src/main/java/dev/speakeasyapi/sdk/utils/SpeakeasyInterceptor.java [hunk 3]
```diff
@@ -14,25 +23,105 @@
...
+
         new Thread(() -> {
             requestResponseHandler.captureRequestResponse((ContentCachingRequestWrapper) req,
-                    (ContentCachingResponseWrapper) res, startTime, new Date());
+                    (ContentCachingResponseWrapper) res, startTime, Instant.now(), pathHint,
+                    controller.getCustomerID());
         }).start();
     }
+
+    private String getPathHint(HandlerMethod hm) {
```
Comment (alexadrake) [inline src/main/java/dev/speakeasyapi/sdk/utils/SpeakeasyInterceptor.java:90]: do we know if these will pull in the path set at the controller level?

File: src/main/java/dev/speakeasyapi/sdk/utils/SpeakeasyInterceptor.java (hunk 2 above)
Comment (TristanSpeakEasy) [reply src/main/java/dev/speakeasyapi/sdk/utils/SpeakeasyInterceptor.java:48]: so these aren't meant to be something really set-able by the end user and mainly for internal use and there are some places where I really think the way we configure things should be consistent between SDKs

File: src/main/java/dev/speakeasyapi/sdk/utils/SpeakeasyInterceptor.java (hunk 3 above)
Comment (TristanSpeakEasy) [reply src/main/java/dev/speakeasyapi/sdk/utils/SpeakeasyInterceptor.java:90]: do you have an example of that? I can setup a test to make sure

File: src/main/java/dev/speakeasyapi/sdk/utils/SpeakeasyInterceptor.java (hunk 2 above)
Comment (TristanSpeakEasy) [reply src/main/java/dev/speakeasyapi/sdk/utils/SpeakeasyInterceptor.java:48]: it also simplifies the code a end user needs to write to support additional instances of the SDK

File: src/main/java/dev/speakeasyapi/sdk/utils/SpeakeasyInterceptor.java (hunk 3 above)
Comment (alexadrake) [reply src/main/java/dev/speakeasyapi/sdk/utils/SpeakeasyInterceptor.java:90]: https://www.baeldung.com/spring-controller-vs-restcontroller you can see in one of the examples here that the controller has a `@RequestMapping` annotation

File: src/main/java/dev/speakeasyapi/sdk/utils/SpeakeasyInterceptor.java (hunk 2 above)
Comment (alexadrake) [reply src/main/java/dev/speakeasyapi/sdk/utils/SpeakeasyInterceptor.java:48]: fair

File: src/main/java/dev/speakeasyapi/sdk/utils/SpeakeasyInterceptor.java (hunk 3 above)
Comment (TristanSpeakEasy) [reply src/main/java/dev/speakeasyapi/sdk/utils/SpeakeasyInterceptor.java:90]: okay it doesn't capture the controller level path so will look into how to ensure that gets captured as well

**Review by simplesagar — COMMENTED** [review]
Just going through this. Could you also add section in the gitbook under the go and java SDKs on using path hints, appID, versionID, and capturing Customer ID as an optional field? Not opposed to just copying the whole readme over there to keep things consistent between the two.

File: src/main/java/dev/speakeasyapi/sdk/utils/SpeakeasyInterceptor.java (hunk 3 above)
Comment (TristanSpeakEasy) [reply src/main/java/dev/speakeasyapi/sdk/utils/SpeakeasyInterceptor.java:90]: @alexadrake fixed

**Comment by TristanSpeakEasy** [comment]
> Just going through this. Could you also add section in the gitbook under the go and java SDKs on using path hints, appID, versionID, and capturing Customer ID as an optional field? Not opposed to just copying the whole readme over there to keep things consistent between the two.

@simplesagar yeah I think someway or aligning the docs with the readme would be good. Ideally automatically one day but yeah I can just copy it over for now?

File: src/main/java/dev/speakeasyapi/sdk/utils/SpeakeasyInterceptor.java (hunk 2 above)
Comment (TristanSpeakEasy) [reply src/main/java/dev/speakeasyapi/sdk/utils/SpeakeasyInterceptor.java:48]: just FYI, Alexa and I discussed and the env vars help with consistent instructions from the helm chart for configuring the SDK for use with that

**Review by alexadrake — APPROVED** [review]

File: README.md [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+    controller.setPathHint("/v1/users/{id}"); // This path hint will be used to match requests to your OpenAPI Schema
+
+    // your handler logic here
+}
+```
+
+## Capturing Customer IDs
+
+To help associate requests with customers/users of your APIs you can provide a customer ID per request handler:
+
```
Comment (simplesagar) [inline README.md:119]: same comment as go SDK. Lets encourage user to set customer id so we can power request viewer filters

**Review by ThomasRooney — APPROVED** [review]
lgtm

**Review by simplesagar — DISMISSED** [review]
No major comments, looks great !

File: src/main/java/dev/speakeasyapi/sdk/SpeakeasyRequestResponseHandler.java [hunk 5]
```diff
@@ -15,8 +15,8 @@
 import javax.servlet.http.HttpServletResponse;
 
 public class SpeakeasyRequestResponseHandler {
-    private final String sdkName = "speakeasy-jvm-sdk";
-    private final String speakeasyVersion = "0.0.1";
+    private final String sdkName = "speakeasy-jvm-springboot-sdk";
+    private final String speakeasyVersion = "1.1.0";
```
Comment (alexadrake) [inline src/main/java/dev/speakeasyapi/sdk/SpeakeasyRequestResponseHandler.java:19]: if you're going to bump the version here, does it make sense to bump the version in the readme and build.gradle? That way I can just create the release without making a new PR to bump the readme version.

**Review by alexadrake — APPROVED** [review]

File: src/main/java/dev/speakeasyapi/sdk/SpeakeasyRequestResponseHandler.java (hunk 5 above)
Comment (TristanSpeakEasy) [reply src/main/java/dev/speakeasyapi/sdk/SpeakeasyRequestResponseHandler.java:19]: yep sorry didn't realize it was defined elsewhere. I imagine this is eventually something we will want driven from a single source of truth

Comment (TristanSpeakEasy) [reply src/main/java/dev/speakeasyapi/sdk/SpeakeasyRequestResponseHandler.java:19]: fixed

**Review by simplesagar — APPROVED** [review]

**Pull request merged**